library/std/src/net/ip.rs

```diff
@@ -1153,10 +1369,18 @@ impl Ipv6Addr {
     /// assert_eq!(Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0xc00a, 0x2ff).is_loopback(), false);
     /// assert_eq!(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0x1).is_loopback(), true);
     /// ```
+    ///
+    /// # Stability guarantees
+    ///
+    /// This method's behavior [may by subject to changes][changes] in the
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
     #[rustc_const_unstable(feature = "const_ipv6", issue = "76205")]
     #[stable(since = "1.7.0", feature = "ip_17")]
     pub const fn is_loopback(&self) -> bool {
         u128::from_be_bytes(self.octets()) == u128::from_be_bytes(Ipv6Addr::LOCALHOST.octets())
+            || if let Some(v4_addr) = self.to_ipv4() { v4_addr.is_loopback() } else { false }
```

Comment:
Regarding the ipv4 in ipv6 concern by @KodrAus:

Using `to_ipv4` causes Rust to consider both ipv4-mapped and ipv4-compatible addresses, i.e. `::ffff:127.0.0.1` and `::127.0.0.1`. This is inconsistent with other languages:

| Language | ipv4-mapped | ipv4-compatible | demo |
| --- | --- | --- | --- |
| Go | ✔️ | ❌  | https://play.golang.org/p/dqDfRtz0AGW |
| C# | ✔️ (since .NET 5)| ❌  | https://dotnetfiddle.net/hhYXgK |
| Java | ✔️ | ❌  | https://ideone.com/2U0peb |

IPv4-Compatible IPv6 address have also been officially deprecated for a long time https://github.com/rust-lang/rust/pull/75150#issuecomment-668680333, therefore I think it would be better to use `to_ipv4_mapped` everywhere.

Reply:
Would the best way forward to create a `to_ipv4_compatible` method that is not called anywhere in checking ipv4-in-ipv6 stuff, or should we just disregard IPv4-compatible addresses entirely?